Add DeadlineInstances

saeltz opened the pull request:
We wanted to have `min` and `max` on a `NonEmptyList`. That requires a `cats.Order[A]`.
For the standard library's `Deadline` no instances were defined. This adds it.

(I'm sure I need to touch some other places to make it available everywhere. Please tell me where. Thanks!)

rossabaker commented:
What's here looks like a good start!

Search for `FiniteDurationInstances`, and I would imagine it'll be in exactly the same places.  I think that will include an `all` package object and maybe a new `deadline` package object for a la carte imports.

See also `FiniteDurationSuite`.  We'll want a similar `DeadlineSuite` to check the laws.

saeltz commented:
@rossabaker, I added the `package object`s, all `trait`s and included this in `all`.

The `FiniteDurationSuite` only tests for the `Show` typeclass which is not available for `Deadline`.
The tests for `FiniteDuration` are in the `OrderSuite` where I added some tests which are passing on JVM but failing on Scala.js. Any hints on how to prevent that?

saeltz commented:
@rossabaker, I fixed all the tests and binary compatibility issues. Please review again. Thanks!

johnynek commented on kernel/src/main/scala/cats/kernel/instances/DeadlineInstances.scala:
```diff
@@ -0,0 +1,34 @@
+  override def maxBound: Deadline = Deadline(FiniteDuration(Long.MaxValue, TimeUnit.NANOSECONDS))
+}
+
+class DeadlineOrder extends Order[Deadline] with Hash[Deadline] with DeadlineBounded { self =>
```
I think a BoundedSemilattice may also make sense here (on max?) or maybe it is unclear so we should let users make their own.

saeltz replied:
It definitely is unclear to me how to implement that. So I'd appreciate either an approval or some help.

rossabaker replied:
The `BoundedSemilattice` is a `Monoid`, so you'd need to define `empty` and `combine`.  You also inherit these typeclasses, and therefore laws:

* a `CommutativeSemigroup`: `combine(x, y) === combine(y, x)`
* a `Band`: `combine(x, x) == x`

If you combine two deadlines, would you want the result to be the earlier deadline (`min`) or the later (`max`)?  I can imagine both, so I'm a bit skeptical of the instance.

And what should `empty` be?  `Deadline.now` is tempting, but changes on each call.  I suppose we'd use `minBound` if we combined with `max`, and `maxBound` if we combined with `min`.  Or we could escape this question by being a `Semilattice` instead of a `BoundedSemilattice`.

I think if someone can make a persuasive case for min or max being the canonical instance, we could handle this in a separate PR.

johnynek replied:
empty would be either the minimum or maximum value depending on which BoundedSemilattice you pick. `now` would not be lawful.

johnynek replied:
Note: we already have a constructor function for picking a max or min semilattice if you give the corresponding empty value. I agree it isn't clear if you want min or max here.

Probably some applications would want one and others the other.

saeltz replied:
If it is dependent on the application, I would refrain from implementing it here and merging this PR.
Later one could implement the `Semilattice` in a follow-up PR anyways.

rossabaker replied:
I'm pretty well convinced the `empty` follows from `combine` and any eventual instance should be monoidal.  But there's no obvious canonical instance, and no reason to block this waiting for one.  Let's get a second approval and get this shipped.

rossabaker reviewed (approved):
:+1:
